dft_organizer/aiida/aiida_links_tree.py

```diff
@@ -0,0 +1,273 @@
+    )
+    SELECT p.input_id, p.output_id, p.depth, p.path,
+           n_in.uuid  AS input_uuid,
+           n_out.uuid AS output_uuid
+    FROM path_up AS p
+    JOIN db_dbnode AS n_in  ON n_in.id  = p.input_id
+    JOIN db_dbnode AS n_out ON n_out.id = p.output_id
+    ORDER BY p.depth, p.input_id ASC;
+    """
+    cur.execute(sql_up, (start_pk,))  #tuple
+    for input_id, output_id, depth, path, in_uuid, out_uuid in cur.fetchall():
+        records.append(
+            LinkRecord(
+                direction="up",
+                depth=depth,
+                input_id=input_id,
+                output_id=output_id,
+                input_uuid=str(in_uuid),
+                output_uuid=str(out_uuid),
+                path=path,
+            )
+        )
+
+    cur.close()
+    return records
+
+
+def find_first_last_pks(conn, start_pk: int) -> Tuple[int, int]:
+    cur = conn.cursor()
+
+    sql_up = """
+    WITH RECURSIVE path_up AS (
+      SELECT link.input_id, link.output_id, 0::INT AS depth
+      FROM db_dblink AS link
+      WHERE link.output_id = %s
+    UNION ALL
+      SELECT newlink.input_id, existing.output_id, existing.depth + 1 AS depth
+      FROM path_up AS existing
+      JOIN db_dblink AS newlink
+        ON newlink.output_id = existing.input_id
+    )
+    SELECT input_id, depth
+    FROM path_up
+    ORDER BY depth DESC
+    LIMIT 1;
+    """
+    cur.execute(sql_up, (start_pk,))  
+    row = cur.fetchone()
+    first_pk = row[0] if row else start_pk
+
+    sql_down = """
+    WITH RECURSIVE path_down AS (
+      SELECT link.input_id, link.output_id, 0::INT AS depth
+      FROM db_dblink AS link
+      WHERE link.input_id = %s
+    UNION ALL
+      SELECT existing.input_id, newlink.output_id, existing.depth + 1 AS depth
+      FROM path_down AS existing
+      JOIN db_dblink AS newlink
+        ON newlink.input_id = existing.output_id
+    )
+    SELECT output_id, depth
+    FROM path_down
+    ORDER BY depth DESC
+    LIMIT 1;
+    """
+    cur.execute(sql_down, (start_pk,))  
+    row = cur.fetchone()
+    last_pk = row[0] if row else start_pk
+
+    cur.close()
+    return first_pk, last_pk
+
+
+def _node_short_info(pk: int):
+    node = load_node(pk)
+    base = f"pk={pk} {node.__class__.__name__}"
+    if hasattr(node, "process_label") and node.process_label:
+        return f"{base} ({node.process_label})"
+    if getattr(node, "label", None):
+        return f"{base} (label='{node.label}')"
+    return base
+
+
+
+def _node_has_structure(node) -> bool:
+    if isinstance(node, StructureData):
+        return True
+    if hasattr(node, "outputs") and "structure" in node.outputs:
+        return isinstance(node.outputs["structure"], StructureData)
+    return False
+
+
+def _collect_structure_nodes(links: List[LinkRecord]) -> List[int]:
+    """Collect all pk nodes where there is a structure (input and output), without duplicates."""
+    pks: set[int] = set()
+    for r in links:
```

Comment:
@alinzh one-letter variables are generally discouraged, except `n` for numbering